tsl/src/fdw/data_node_scan_plan.c

```diff
@@ -674,6 +688,689 @@ push_down_group_bys(PlannerInfo *root, RelOptInfo *hyper_rel, Hyperspace *hs,
 	}
 }
 
+/*
+ * Check if the join fromList belongs to a INNER JOIN / LEFT JOIN
+ * on two tables. A INNER JOIN / LEFT JOIN on two tables has one JoinExpr
+ * entry in fromList.
+ */
+static bool
+is_inner_or_left_join_two_tables(List *fromList)
+{
+	if (!IsA(list_nth(fromList, 0), JoinExpr) || (list_length(fromList) != 1))
+		return false;
+
+	return true;
+}
+
+/*
+ * Check if the join fromList belongs to an implicit join (e.g.,
+ * SELECT * FROM metric m1, metric_name m2 WHERE m1.id=m2.id). This type
+ * of join has two RANGETBLREF entries in the fromList.
+ */
+static bool
+is_implicit_join_two_tables(List *fromList)
+{
+	if (list_length(fromList) != 2)
+		return false;
+
+	/* At the moment, the code is called only in the reference table join
+	 * pushdown path. Earlier checks ensure that both parameters are
+	 * RangeTblRefs. Recheck the condition here and handle the case properly.
+	 */
+	if (!IsA(list_nth(fromList, 0), RangeTblRef))
+		return false;
+
+	if (!IsA(list_nth(fromList, 1), RangeTblRef))
+		return false;
+
+	return true;
+}
+
+/*
+ * Check if the query performs a join between a hypertble (outer) and a reference
+ * table (inner) and the join type is a LEFT JOIN, an INNER JOIN, or an implicit
+ * join.
+ */
+static bool
+is_safe_to_pushdown_reftable_join(PlannerInfo *root, List *join_reference_tables,
+								  RangeTblEntry *outertableref, RangeTblEntry *innertableref,
```

Comment:
`outertableref` not used in this function except for asserts. I guess it is maybe because we don't support all kinds of joins (yet), but consider eliding this parameter from the parameter list.

Reply:
Good point, the parameter is removed.